doc/developer/design/20230810_alter_connection.md

```diff
@@ -0,0 +1,64 @@
+# `ALTER CONNECTION`
+
+- Associated:
+    - https://github.com/MaterializeInc/materialize/issues/17620
+
+## Context
+
+A `CONNECTION` must be valid for a source to ingest data--however, in our
+current scheme, `CONNECTION`s are immutable so any change to the service hosting
+the source permanently wedges the source in Materialize. This is undesirable
+because the `DROP`/`CREATE` cycle for a source can be very painful for users
+because of rehydration.
+
+## Goals
+
+Allow users to modify a `CONNECTION` s/t they can repair the TCP connection
+between MZ and their services.
+
+## Non-Goals
+
+Allow users to meticulously control their sources' state changes.
+
+## Overview
+
+We can perform a process similar to `ALTER SOURCE` wherein we update some state
+in memory, re-plan something, update it in the catalog, and then communicate the
+updated state throughout the system.
+
+## Detailed description
+
+1. Parse `ALTER CONNECTION <name> AS (KAFKA|POSTGRES) SET <option> =? <val>`.
+    - In parsing `AS` clause dictates the options available. These will get
+      unified in an enum that passes from parsing into planning.
+1. During planning/purification:
+    1. Check if `AS` clause matches existing connection type.
+    1. Perform connection validation using same logic as `CREATE CONNECTION`.
+1. During sequencing:
+    1. Load the connection in memory. Update its values.
+    1. Re-plan the value to get its updated `CREATE CONNECTION` string, just as
+       we do for `ALTER SOURCE`.
+    1. Commit a catalog update op for the connection.
+    1. Find all sources that use the connection and update their ingestion
+       description to use the new connection. I believe this only needs to be
+       done to the in-memory version of the `SOURCE` because the on-disk
+       representation only contains the connection details by name.
+    1. To the storage controller, send a `RunIngestion` command with the new,
+       in-memory representation of the ingestion.
+
+       This will restart all of the sources with the new connection state.
```

Comment:
I'd be interested in slicing this work off and making it part of another deliverable. `ALTER SECRET` already has the issue that altering a secret does not automatically propagate the new secret value to already running sources and sinks. But there's a workaround! You can drop and recreate the _replica_ of the source cluster—and then bam, updated sources. We recently updated the `ALTER SECRET` docs to describe this workaround: https://materialize.com/docs/sql/alter-secret/

Obviously this sucks and we should fix it, but I think it's totally okay to ship v1 of this without support for auto-restart, and then we can ship v2 with auto restart for _both_ `ALTER SECRET` and `ALTER SOURCE`.

Reply:
We chatted about this offline and auto-restart on `ALTER` for `CONNECTION` should be trivial to implement. I believe we can also do this trivially for `ALTER SECRET`, but need to check some code before committing to it.